include/deal.II/lac/lapack_full_matrix.h

```diff
@@ -420,6 +420,42 @@ class LAPACKFullMatrix : public TransposeTable<number>
                const LAPACKFullMatrix<number> &B,
                const bool                      adding=false) const;
 
+  /**
+   * Matrix-matrix-multiplication using transpose of <tt>this</tt> and a
+   * diagonal vector @p V.
+   *
+   * If the <code>adding=false</code> then the result is stored in the matrix
+   * $C = A^T \rm{diag}(V) B$
+   * otherwise it is added $C \mathrel{+}= A^T \rm{diag}(V) B$.
+   *
+   * @note It is assumed that @p A, @p B and @p V have compatible sizes and that
+   * @p C already has the right size.
+   *
+   * @note This function is not provided by LAPACK. The function first forms $BV$ product and
+   * then uses Xgemm function.
+   */
+  void Tmmult (LAPACKFullMatrix<number>       &C,
+               const LAPACKFullMatrix<number> &B,
+               const Vector<number>           &V,
+               const bool                      adding=false) const;
+
+  /**
+   * Matrix-matrix-multiplication using transpose of <tt>this</tt> and a
+   * diagonal vector @p V. Similar to the above but computes $C = A^T \rm{diag}(V)^2 A$
+   * when <code>adding=false</code> and $C \mathrel{+}= A^T \rm{diag}(V)^2 A$
+   * otherwise. If @p V is empty, it's taken as identity matrix which corresponds
+   * to calculation of $A^T A$.
+   *
+   * @note It is assumed that @p A and @p V have compatible sizes and that
+   * @p C already has the right size.
+   *
+   * @note This function is not provided by LAPACK. The function first forms $VA$ product and
+   * then uses Xsyrk function.
+   */
+  void Tmmult (LAPACKFullMatrix<number>       &C,
+               const Vector<number>           &V = Vector<number>(),
+               const bool                      adding=false) const;
```

Comment:
I am confused about why you introduce two functions. One can always call the former with setting `B=A` and get the same result. Is there anything I miss? I'm a bit reluctant on making the interfaces overly big without a good reason.

As to the possible optimization by using `dsyrk` instead of `dgemm`, couldn't we achieve this internally if we detect that `B.data()` points to the same array as `A.data()`?

Reply:
@kronbichler 

>  One can always call the former with setting B=A and get the same result.

Nope, the two are different, one is `A^T * V * V * A` and another is `A^T * V * B`, that is, the difference is in `V^2` vs `V`.

Reply:
> As to the possible optimization by using dsyrk instead of dgemm, couldn't we achieve this internally if we detect that B.data() points to the same array as A.data()?

that's a question to `Tmmult` without a vector argument. There we could indeed go with `dsyrk` instead of `dgemm`.

Reply:
> Nope, the two are different, one is `A^T * V * V * A` and another is `A^T * V * B`.

I see. But then the question is whether we need that or whether we have that somewhere else. I remember that we have the latter, but I haven't seen the former. As I said, I'm a bit concerned of the size of the API if we include cases which are easily translated into one another in user code, in this case with O(n) operations. But if others don't see this as a concern, I'm also perfectly fine on adding this second function to the interface.

Reply:
> whether we need that

I add this because I may use it in my own code. 

>  we have that somewhere else.

I don't think this operation is currently implemented in other linear algebras.

> we include cases which are easily translated into one another in user code

that's not the case, at the moment. There is nothing in LAPACKFullMatrix that can help with `A^T V V A`.

Reply:
> that's not the case, at the moment. There is nothing in LAPACKFullMatrix that can help with `A^T V V A`.

You simply square the entries in the vector `V` and pass that into the other function that you add with this PR, don't you?

Reply:
> You simply square the entries in the vector V and pass that into the other function that you add with this PR, don't you?

oh, but that is completely different code and slower and won't take into account symmetry of the resulting `C`. I mean there is nothing that can help using `dsyrk` in this case.

Reply:
Ah, now I see, yes, you can't use `dsyrk`, that's true, but with a high-performance BLAS implementation its impact should only be visible above `100x100` matrices or something like that. But yes, it is an argument that we could think of adding it. Any other opinions?

Reply:
> Any other opinions?

@dealii/dealii ping. 

I can remove the function in question and achieve the same efficiency by adding manually written `LAPACKFullMatrix::scale_rows(const Vector<number> &V)` and using `Tmmult()` which is updated in this PR as well.

Reply:
nevermind, i will remove the function and add `scale_rows()` to avoid confusion in the interface.